src/OpenSage.Game/Gui/Apt/ShapeComponent.cs

```diff
@@ -0,0 +1,218 @@
+        {
+            // Figure out the ratio.
+            var ratioX = viewportSize.Width / (float) shapeBounding.Width;
+            var ratioY = viewportSize.Height / (float) shapeBounding.Height;
+
+            // Use whichever multiplier is smaller.
+            var ratio = ratioX < ratioY ? ratioX : ratioY;
+
+            scale = ratio;
+
+            var originalWidth = shapeBounding.Width;
+            var originalHeight = shapeBounding.Height;
+
+            // Now we can get the new height and width
+            var newWidth = (int) Math.Round(originalWidth * ratio);
+            var newHeight = (int) Math.Round(originalHeight * ratio);
+
+            newWidth = Math.Max(newWidth, 1);
+            newHeight = Math.Max(newHeight, 1);
+
+            var newX = (int) Math.Round(shapeBounding.X * ratio);
+            var newY = (int) Math.Round(shapeBounding.Y * ratio);
+
+            // Now calculate the X,Y position of the upper-left corner 
+            // (one of these will always be zero for the top level window)
+            var posX = (int) Math.Round((viewportSize.Width - (shapeBounding.Width * ratio)) / 2);
+            var posY = (int) Math.Round((viewportSize.Height - (shapeBounding.Height * ratio)) / 2);
+
+            return new Rectangle(posX, posY, newWidth, newHeight);
+        }
+
+        public void Layout(GraphicsDevice gd, in Size windowSize)
+        {
+            float _scale = 0.0f;
+
+            var Frame = CalculateFrame(Shape.BoundingBox, windowSize, out _scale);
+
+            _texture = Texture.CreateTexture2D(
+                gd,
+                PixelFormat.Rgba8UNorm,
+                Frame.Width,
+                Frame.Height,
+                TextureBindFlags.ShaderResource | TextureBindFlags.RenderTarget);
+
+            Material.SetTexture(_texture);
+
+            var left = (Frame.X / (float) windowSize.Width) * 2 - 1;
+            var top = (Frame.Y / (float) windowSize.Height) * 2 - 1;
+            var right = ((Frame.X + Frame.Width) / (float) windowSize.Width) * 2 - 1;
+            var bottom = ((Frame.Y + Frame.Height) / (float) windowSize.Height) * 2 - 1;
+           
+            var vertices = new[]
+            {
+                new SpriteVertex(new Vector2(left, top * -1), new Vector2(0, 0)),
+                new SpriteVertex(new Vector2(right, top * -1), new Vector2(1, 0)),
+                new SpriteVertex(new Vector2(left, bottom * -1), new Vector2(0, 1)),
+                new SpriteVertex(new Vector2(right, top * -1), new Vector2(1, 0)),
+                new SpriteVertex(new Vector2(right, bottom * -1), new Vector2(1, 1)),
+                new SpriteVertex(new Vector2(left, bottom * -1), new Vector2(0, 1))
+            };
+            _vertexBuffer = Buffer<SpriteVertex>.CreateStatic(
+                gd,
+                vertices,
+                BufferBindFlags.VertexBuffer);
+
+        }
+
+        private ColorRgbaF ToFloatColor(ColorRgba color)
```

Comment:
This already exists, as the incorrectly-named `ColorRgba.ToColorRgba()` method. Feel free to use that method and fix its name :)